File: content/problems/3134-find-the-median-of-the-uniqueness-array/solutions/binary-search-distinct-limit-with-window/Solution.kt
class Solution {
    private fun countAtMost(nums:IntArray,limit:Int):Long{
        val frequency=IntArray(100001);
        var left=0;
        var distinct=0;
        var count=0L;
        for(right in nums.indices){
            if(frequency[nums[right]]++==0)distinct++;
            while(distinct>limit){
                if(--frequency[nums[left++]]==0)distinct--
            };
            count+=right-left+1
        };
        return count
    }
    fun medianOfUniquenessArray(nums:IntArray):Int{
        val total=nums.size.toLong()*(nums.size+1)/2;
        val target=(total+1)/2;
        var low=1;
        var high=nums.size;
        while(low<high){
            val middle=(low+high)/2;
            if(countAtMost(nums,middle)>=target)high=middle else low=middle+1
        };
        return low
    }
}

File: content/problems/3134-find-the-median-of-the-uniqueness-array/solutions/binary-search-distinct-limit-with-window/solution.cpp
class Solution {
    long long countAtMost(vector<int>& nums,int limit){
        vector<int> frequency(100001);
        int left=0,distinct=0;
        long long count=0;
        for(int right=0;right<(int)nums.size();++right){
            if(frequency[nums[right]]++==0)++distinct;
            while(distinct>limit)if(--frequency[nums[left++]]==0)--distinct;
            count+=right-left+1;
        }
        return count;
    }
    public:int medianOfUniquenessArray(vector<int>& nums){
        long long total=1LL*nums.size()*(nums.size()+1)/2,target=(total+1)/2;
        int low=1,high=nums.size();
        while(low<high){
            int middle=(low+high)/2;
            if(countAtMost(nums,middle)>=target)high=middle;
            else low=middle+1;
        }
        return low;
    }
};
